cirq/google/arg_func_langs.py

```diff
@@ -12,9 +12,196 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import List, Union
+from typing import List, Union, Optional, Iterator, Iterable, cast, Set
 
+import numpy as np
 import sympy
+from cirq.api.google import v2
 
-# Argument types for gates. ArgFunction's are not currently supported.
-ArgValue = Union[int, float, List[bool], str, sympy.Symbol]
+SUPPORTED_FUNCTIONS_FOR_LANGUAGE = {
+    '': set(),
+    'linear': {'add', 'mul'},
```

Comment:
A general question:  you've made things very general here.  However, I am trying to think through why we might need things so complicated.  This is all in cirq/google, so it can be specific to our use case.  Can you explain a case why we might need multiple different arg languages besides nothing and everything?

Is this to provide a smooth upgrade path for when we add new functions to servers/clients at different times?  I suppose the simulator could support less options?

Reply:
It's to provide a smooth upgrade path. For example, suppose we add support for trig functions. Clients that support trig functions but run parameterized circuits that don't use trig but do use linear combinations will still be able to talk to a server that only supports linear combinations.